File: src/main/scala/robosky/ether/block/unbreakable/Unbreakable.scala
package robosky.ether.block.unbreakable

import com.google.common.collect.ImmutableMap

import net.minecraft.block.{Block => McBlock, BlockState}
import net.minecraft.state.property.Property
import net.minecraft.util.math.BlockPos
import net.minecraft.world.BlockView

/**
 * A mixin trait for unbreakable blocks.
 */
trait Unbreakable extends McBlock {

  /**
   * The block this ubreakable block is imitating.
   */
  val base: McBlock

  /**
   * Returns the breakable state equivalent of this unbreakable state.
   */
  def toBreakable(state: BlockState): BlockState = {
    import scala.language.existentials
    type Entries = ImmutableMap[Property[T], T] forSome { type T <: Comparable[T] }
    val propValues = state.getEntries.asInstanceOf[Entries]
    var res = base.getDefaultState
    propValues forEach { (p, v) => res = res.`with`(p, v) }
    res
  }

  override def getHardness(state: BlockState, world: BlockView, pos: BlockPos): Float = -1.0f

  override def getBlastResistance: Float = 3600000.0f
}

File: src/main/java/robosky/uplands/block/unbreakable/Unbreakable.java
package robosky.uplands.block.unbreakable;

import java.util.Map;

import com.google.common.collect.ImmutableMap;

import net.minecraft.block.Block;
import net.minecraft.block.BlockState;
import net.minecraft.state.property.Property;

/**
 * An interface that denotes a block as unbreakable. Unbreakable blocks in a boss room
 * are converted to equivalent breakable blocks when a boss is defeated.
 */
public interface Unbreakable {

    /**
     * The breakable version of this block.
     */
    Block base();

    /**
     * Returns the breakable state equivalent of this unbreakable state.
     */
    default BlockState toBreakable(BlockState state) {
        ImmutableMap<Property<?>, Comparable<?>> propValues = state.getEntries();
        BlockState res = this.base().getDefaultState();
        for (Map.Entry<Property<?>, Comparable<?>> entry : propValues.entrySet()) {
            res = with(res, entry.getKey(), entry.getValue());
        }
        return res;
    }

    // public because interface, static because interface static methods are not inherited
    // make private when Minecraft starts using Java 9+
    static <T extends Comparable<T>> BlockState with(BlockState res, Property<T> key, Comparable<?> value) {
        return res.with(key, key.getType().cast(value));
    }
}
